src/mapping/query/function/slice.rs

```diff
@@ -0,0 +1,269 @@
+use super::prelude::*;
+
+#[derive(Debug)]
+pub(in crate::mapping) struct SliceFn {
+    query: Box<dyn Function>,
+    start: Box<dyn Function>,
+    end: Option<Box<dyn Function>>,
+}
+
+impl SliceFn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(
+        query: Box<dyn Function>,
+        start: isize,
+        end: Option<isize>,
+    ) -> Self {
+        let start = Box::new(Literal::from(Value::from(start as i64)));
+        let end = end.map(|i| Box::new(Literal::from(Value::from(i as i64))) as _);
+
+        Self { query, start, end }
+    }
+}
+
+impl Function for SliceFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let mut start = match self.start.execute(ctx)? {
+            Value::Integer(v) => v,
+            v => unexpected_type!(v),
+        };
+
+        let end = &self
+            .end
+            .as_ref()
+            .map(|v| v.execute(ctx))
+            .transpose()?
+            .map(|v| match v {
+                Value::Integer(v) => v,
+                v => unexpected_type!(v),
+            });
+
+        let mut range = |len: i64| {
+            if start < 0 {
+                start += len;
+            }
+
+            let mut end = match end {
+                Some(end) => *end,
+                None => len,
+            };
+
+            if end < 0 {
+                end += len;
+            }
```

Comment:
On the `let mut` hunt again:
```suggestion
            let end = match end {
                Some(&end) if end < 0 => end + len,
                Some(&end) => end,
                None => len,
            };
```